packages/chakra-components/src/components/Election/Election.tsx

```diff
@@ -80,21 +91,72 @@ export const useElectionProvider = ({
   useEffect(() => {
     if (!fetchCensus || !signer || !election || !loaded || !client || isAbleToVote !== undefined) return
     ;(async () => {
+      const censusType: CensusType = election.census.type
       const isIn = await client.isInCensus(election.id)
       let left = 0
-      if (isIn) {
+      if (isIn || censusType == CensusType.WEIGHTED) {
         // no need to check votes left if member ain't in census
         left = await client.votesLeftCount(election.id)
         setVotesLeft(left)
 
         const voted = await client.hasAlreadyVoted(election.id)
         setVoted(voted)
       }
+      setCensusType(censusType)
       setIsInCensus(isIn)
-      setIsAbleToVote(left > 0 && isIn)
+      setIsAbleToVote((left > 0 && isIn) || censusType == CensusType.CSP)
     })()
   }, [fetchCensus, election, loaded, client, isAbleToVote, signer])
 
+  // CSP OAuth flow
+  // Listening for the popup window meessage (oauth flows)
+  useEffect(() => {
+    ;(async () => {
+      const handleMessage = (event: any) => {
+        if (event.data.code && event.data.handler) {
+          getOAuthToken(client, event.data.code, event.data.handler)
+        }
+      }
+
+      if (window.opener || !client || censusType !== CensusType.CSP) {
+        return
+      }
+
+      window.addEventListener('message', handleMessage)
+
+      return () => {
+        window.removeEventListener('message', handleMessage)
+      }
+    })()
+  }, [client, censusType])
+
+  // CSP OAuth flow
+  // Posting the message to the main window
+  useEffect(() => {
+    ;(async () => {
+      if (typeof window == 'undefined') return
+      if (window.location.href.split('?').length < 2) return
+
+      const code = window.location.href
+        .split('?')[1]
+        .split('&')
+        .find((param: string) => param.startsWith('code='))
+        ?.split('=')[1]
+      const handler = window.location.href
+        .split('?')[1]
+        .split('&')
+        .find((param: string) => param.startsWith('handler='))
+        ?.split('=')[1]
```

Comment:
I think you should use the URLSearchParams API here: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams